rpgJavaInterpreter-core/src/main/kotlin/com/smeup/rpgparser/interpreter/internal_interpreter.kt

```diff
@@ -382,6 +382,22 @@ class InternalInterpreter(val systemInterface: SystemInterface) {
                     }
                     assign(statement.target, value.negate())
                 }
+                is SubStmt -> {
+                    val minuend = if (statement.factor1 == null) {
+                        interpret(statement.target)
+                    } else {
+                        interpret(statement.factor1)
+                    }
+                    require(minuend is NumberValue) {
+                        "$minuend should be a number"
+                    }
+                    val subtrahend = interpret(statement.factor2)
+                    require(subtrahend is NumberValue) {
+                        "$subtrahend should be a number"
+                    }
+                    val newValue:DecimalValue = DecimalValue(minuend.asDecimal().value.subtract(subtrahend.asDecimal().value))
```

Comment:
it is not necessary to specify the type of newValue, as it can be inferred.

Regarding the value, I am not sure we always want to get a DecimalValue. In most languages when subtracting two integers I obtain an integer. I obtain a decimal value when at least one of the values is a decimal.
This is important because if the target is an integer trying to assign a DecimalValue should trigger an error (unless coerce fix that for us, but we need to be sure this is the behavior we want)

Reply:
How can i do that? the Value class has not a subtract method. I had to cast minuend. Is there another way?

Reply:
Using the `is` operator. Look at what we do for the PlusExpr, for example:

```
            is PlusExpr -> {
                val left = interpret(expression.left)
                val right = interpret(expression.right)
                when {
                    left is StringValue && right is StringValue -> {
                        val s = left.valueWithoutPadding + right.valueWithoutPadding
                        StringValue(s)
                    }
                    left is IntValue && right is IntValue -> IntValue(left.value + right.value)
                    else -> throw UnsupportedOperationException("I do not know how to sum $left and $right at ${expression.position}")
                }
            }
```

Also, in this way you can check we are subtracting things with the proper type. We do not want to allow subtracting two strings, for example.

Reply:
Also, if you want you could add this logic in a subtract method for Value

Reply:
implemented with when